torch/lib/c10d/reducer.cpp

```diff
@@ -673,9 +669,31 @@ void Reducer::mark_variable_ready(VariableIndex index) {
     if (find_unused_parameters_) {
       // H2D from local_used_maps_ to local_used_maps_dev_
       for (size_t i = 0; i < local_used_maps_.size(); i++) {
-        // We do async H2D to avoid the blocking overhead. The async copy and
-        // allreduce respect the current stream, so will be sequenced correctly.
-        local_used_maps_dev_[i].copy_(local_used_maps_[i], true);
+        if (local_used_maps_dev_[i].is_cuda()) {
+          // Note [local_used_maps_ -> local_used_maps_dev copying]
+          // ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+          // We do async H2D to avoid the blocking overhead. The async copy and
+          // allreduce respect the current stream, so will be sequenced correctly.
+          //
+          // Correct sequencing with respect to host operations is also essential.
+          // The H2D copy_ is stream ordered, while the host's changes to local_used_maps_
+          // are host ordered. If a large backlog of cuda-stream work pushes the copy_ far
+          // into the future, and if no blocking calls occur between now and finalize_backward()**
+          // such that finalize_backward() re-zeroes local_used_maps_ on the host
+          // before the stream executes the copy_, copy_ will read those zeros instead of the
+          // values we thought we told it to read here.
+          // Copying local_used_maps_[i] to a pinned temporary (which the pinned caching
+          // allocator should supply asynchronously) avoids this nasty, rare race condition.
+          //
+          // ** In the hoped-for case where all params are used, DDP itself won't do any
+          // blocking work between now and the re-zeroing, so the danger is real.
+          auto local_used_maps_tmp = local_used_maps_[i].pin_memory();
```

Comment:
If it were a raw cuda pinned memory request (ie cudaHostAlloc or cudaMallocHost) every time, then yes, it would be a performance concern. But THC uses a fast caching allocator for host-pinned memory (separate from the caching allocator used for device memory), and this call only happens once for the entire backward pass, so I think we're alright.

Of course we should stay alert what happens to find_unused_parameters performance after this change. Do you have any particular models/benchmarks in mind? We could try some of them on our end.

Reply:
Hey @zhaojuanmao, I recall you worked on ResNet and RoBERTa perf test before. Any suggestions on which OSS models should be used to benchmark this change?

Reply:
@mcarilli and @mrshenli I can benchmark this diff with Bert and Resnet model later on.